kubeflow/trainer/rhai/transformers.py

```diff
@@ -22,15 +63,452 @@ class TransformersTrainer:
+            )
+
+
+def _create_progression_instrumentation(metrics_port: int):
+    """Instrumentation code injected into training pods (extracted via inspect.getsource).
+
+    This function is NOT called directly in the SDK - it's extracted as source code
+    via inspect.getsource() and injected into user training scripts. This approach
+    provides syntax highlighting, testability, and type checking while avoiding
+    runtime SDK dependencies.
+
+    Args:
+        metrics_port: Port for HTTP metrics server
+
+    Returns:
+        Tuple of (apply_fn, callback_class, handler_class) for testing purposes
+    """
+    from dataclasses import asdict, dataclass, field
+    import http.server
+    import json
+    import threading
+    import time
+    from typing import Any, Optional
+
+    from transformers import TrainerCallback, trainer as trainer_module
+
+    @dataclass
+    class ProgressionMetricsState:
+        """Progression metrics state (camelCase for Kubernetes API compatibility)."""
+
+        progressPercentage: Optional[int] = None  # noqa: N815
+        estimatedRemainingSeconds: Optional[int] = None  # noqa: N815
+        currentStep: int = 0  # noqa: N815
+        totalSteps: Optional[int] = None  # noqa: N815
+        currentEpoch: int = 0  # noqa: N815
+        totalEpochs: Optional[int] = None  # noqa: N815
+        trainMetrics: dict[str, Any] = field(default_factory=dict)  # noqa: N815
+        evalMetrics: dict[str, Any] = field(default_factory=dict)  # noqa: N815
+
+    _progression_metrics_state = ProgressionMetricsState()
+    _progression_metrics_lock = threading.Lock()
+
+    def _update_progression_metrics(updates: dict) -> None:
+        """Thread-safe metrics update."""
+        with _progression_metrics_lock:
+            for key, value in updates.items():
+                if hasattr(_progression_metrics_state, key):
+                    current_value = getattr(_progression_metrics_state, key)
+                    if isinstance(value, dict) and isinstance(current_value, dict):
+                        current_value.update(value)
+                    else:
+                        setattr(_progression_metrics_state, key, value)
+
+    def _get_progression_metrics_json() -> str:
+        """Get current metrics as JSON string."""
+        with _progression_metrics_lock:
+            return json.dumps(asdict(_progression_metrics_state))
+
+    class ProgressionMetricsHandler(http.server.BaseHTTPRequestHandler):
+        """HTTP server that exposes training progress metrics as JSON."""
+
+        def log_message(self, format, *args):
+            """Suppress HTTP server logging."""
+            pass
+
+        def do_GET(self):
+            """Handle GET requests to expose metrics as JSON."""
+            self.send_response(200)
+            self.send_header("Content-Type", "application/json")
+            self.end_headers()
+            self.wfile.write(_get_progression_metrics_json().encode("utf-8"))
```

Comment:
sorry - one small suggestion here for better error reporting


```suggestion
            try: 
                payload = _get_progression_metrics_json()
            except Exception e:
                print(f"[Kubeflow] Failed to create progress metrics payload: {e}")
                self.send_error(500)
            else:
                self.send_response(200)
                self.send_header("Content-Type", "application/json")
                self.end_headers()
                self.wfile.write(payload.encode("utf-8"))
```

Reply:
thanks @robert-bell !
Let me update it right away 👀

Reply:
Done ✅